### DISCUSSION TITLE
If possible, do not show empty spaces

### DISCUSSION THREAD

**Discussion by alexdev16** [YaLTeR/niri#637] [General]
When navigating by columns, do not show empty spaces if all windows in the workspace are within the monitor borders. (for example: create three windows, the first with the maximum width and the other two on the second workspace with a width 50%. If you use gestures, you will see the difference in navigation between the first and second workspace)

**Comment by YaLTeR** [comment]
Hmm, not sure what you mean, could you record a video perhaps?

**Comment by alexdev16** [comment]
Uploading out.mp4…

**Comment by alexdev16** [comment]
https://github.com/user-attachments/assets/71f27326-9d9f-4f69-bc26-3bcf68ea18a0
I would like the second workspace to have the same behavior as the first one

**Reply by YaLTeR** [reply]
Ah, I see. The reason it behaves this way is that sometimes it's useful to have one window to the side. Though, I'm not 100% set on that behavior.

**Reply by alexdev16** [reply]
Strange decision. I don't see any benefit in it.

**Reply by alexdev16** [reply]
There has already been a discussion on this issue, so I'm not alone in my suffering: https://github.com/YaLTeR/niri/discussions/466
I would really like to see this behavior fixed globally or via configuration.

**Reply by YaLTeR** [reply]
I just pushed 952916fd1c61ebd61e7b40211d97fdd0c15e85d1 which prevents this for the gesture. Please try it and tell me how it feels

**Reply by alexdev16** [reply]
Wow, how quickly!  Unfortunately, I will only be able to try in 2 weeks, when my vacation ends

**Comment by alexdev16** [comment]
Thanks for the changes to gesture navigation, everything works as it should. And overall the release is great!

**Reply by alexdev16** [reply]
I noticed a small flaw:
1. create 3 windows with a column width of 50%
2. move focus to the previous window
3. move focus back to the last window
4. close the last window
The workspace is not aligned with the last window.

https://github.com/user-attachments/assets/0967b5e6-c1b2-4715-8fe6-83023c553571

**Reply by YaLTeR** [reply]
This is intentional. When you change focus, the "previous view position" is cleared to avoid surprising behavior. Afterward, it's the default niri behavior to scroll only as far as necessary to see the focused window.

**Reply by chcore** [reply]
> This is intentional. When you change focus, the "previous view position" is cleared to avoid surprising behavior. Afterward, it's the default niri behavior to scroll only as far as necessary to see the focused window.

It would be really useful to have an option allowing niri to utilize the remaining space automatically (so in the video, at the end it would automatically scroll to the left by 50% of the screen width rather than the user doing it).

For me, the current behavior results in some unnecessary keystrokes (manually scrolling left), and is currently surprising behavior from my perspective, coming from other tiling window managers that tend to always fill up the screen.
